koap-core/src/commonMain/kotlin/Message.kt

```diff
@@ -356,6 +382,93 @@ sealed class Message {
                 }
             }
         }
+
+        companion object {
+            fun numMSzx(blockNumber: Long, more: Boolean, blockSize: Int): Long =
+                (blockNumber shl 4) or (if (more) 0x08 else 0) or szx(blockSize)
+
+            fun szx(blockSize: Int): Long = when (blockSize) {
+                16 -> 0
+                32 -> 1
+                64 -> 2
+                128 -> 3
+                256 -> 4
+                512 -> 5
+                1024 -> 6
+                -1 -> 7 // use blockSize -1 for BERT blocks
+                // https://datatracker.ietf.org/doc/html/rfc8323#section-6
+                // https://datatracker.ietf.org/doc/html/draft-bormann-core-block-bert-00#section-2
+                else -> throw IllegalArgumentException("Invalid block size: $blockSize")
+            }
+
+            fun blockNumber(numMSzx: Long) = numMSzx shr 4
+            fun more(numMSzx: Long) = (numMSzx and 0x08) != 0L
+            fun blockSize(numMSzx: Long) = if ((numMSzx and 0x07) != 7L) 16 shl (numMSzx and 0x07).toInt() else -1
+        }
+
+        /** RFC 7959 2.1. The Block2 and Block1 Options */
+        data class Block1(
+            val blockNumber: Long,
+            val more: Boolean,
+            val blockSize: Int,
+        ) : Option() {
+            init {
+                val numMSzx = numMSzx(blockNumber, more, blockSize)
+                require(numMSzx in BLOCK_RANGE) {
+                    "Uri-Port value of $numMSzx is outside allowable range of $BLOCK_RANGE"
+                }
```

Comment:
started with data classes with a val numMSzx, but that seems error prone and hard to use.

not sure if the companion object should be where it is, or part of Message or its own class?

Reply:
I squashed the old version away, since the separate blockNumber, more flag, blockSize seems much more friendly

> not sure if the companion object should be where it is, or part of Message or its own class?

especially since you can only have one companion object block, I guess it should be at the start (or end)?

should the methods be called something with "block", or should it be moved to some block utility companion object somewhere else?